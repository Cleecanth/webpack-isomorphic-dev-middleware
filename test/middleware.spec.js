'use strict';

const fs = require('fs');
const express = require('express');
const request = require('supertest');
const webpackIsomorphicDevMiddleware = require('../');
const createCompiler = require('./util/createCompiler');
const normalizeHtmlError = require('./util/normalizeHtmlError');
const configClientBasic = require('./configs/client-basic');
const configServerBasic = require('./configs/server-basic');
const configClientSyntaxError = require('./configs/client-syntax-error');
const configServerRuntimeError = require('./configs/server-runtime-error');

describe('middleware', () => {
    afterEach(() => createCompiler.teardown());

    it('should wait for a successful compilation and call next()', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerBasic);

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
        }));

        return request(app)
        .get('/client.js')
        .expect(200)
        .expect(/Hello!/);
    });

    it('should wait for a failed compilation and render the webpack stats', () => {
        const app = express();
        const compiler = createCompiler(configClientSyntaxError, configServerBasic);

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
        }));

        return request(app)
        .get('/client.js')
        .expect(500)
        .expect((res) => {
            expect(normalizeHtmlError(res.text)).toMatchSnapshot();
        });
    });

    it('should render error if an error occurred while reading the server bundle', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerRuntimeError);

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
        }));

        compiler.server.webpackCompiler.outputFileSystem.readFile = (...args) => {
            args[args.length - 1](new Error('Failed to read file'));
        };
        compiler.server.webpackCompiler.outputFileSystem.readFileSync = () => {
            throw new Error('Failed to read file');
        };

        return request(app)
        .get('/client.js')
        .expect(500)
        .expect((res) => {
            expect(normalizeHtmlError(res.text)).toMatchSnapshot();
        });
    });

    it('should render error if i the server bundle has a runtime error', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerRuntimeError);

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
        }));

        return request(app)
        .get('/client.js')
        .expect(500)
        .expect((res) => {
            expect(normalizeHtmlError(res.text)).toMatchSnapshot();
        });
    });

    it('should call next(err) if not a middleware error', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerBasic);
        const contrivedError = new Error('foo');

        app.use((req, res, next) => {
            next(contrivedError);
        });
        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
        }));
        app.use((err, req, res, next) => {  // eslint-disable-line handle-callback-err, no-unused-vars
            expect(err).toBe(contrivedError);
            res.send(`Captured contrived error: ${err.message}`);
        });

        return request(app)
        .get('/client.js')
        .expect(200)
        .expect('Captured contrived error: foo');
    });

    it('should set res.locals.isomorphicCompilation', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerBasic);
        let isomorphicCompilation;

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
        }));

        app.get('*', (req, res) => {
            isomorphicCompilation = res.locals.isomorphicCompilation;
            res.send('Yes it works!');
        });

        return request(app)
        .get('/')
        .expect(200)
        .expect('Yes it works!')
        .then(() => {
            expect(isomorphicCompilation).toBeDefined();
            expect(isomorphicCompilation).toHaveProperty('stats.client');
            expect(isomorphicCompilation).toHaveProperty('stats.server');
            expect(isomorphicCompilation).toHaveProperty('exports.render');
        });
    });

    it('should not use in-memory filesystem if options.memoryFs = false', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerBasic);

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            memoryFs: false,
            watchOptions: { report: false },
        }));

        return request(app)
        .get('/client.js')
        .expect(200)
        .then(() => {
            expect(fs.existsSync(`${compiler.client.webpackConfig.output.path}/client.js`)).toBe(true);
            expect(fs.existsSync(`${compiler.server.webpackConfig.output.path}/server.js`)).toBe(true);
        });
    });

    it('should not call watch() automatically if options.watchOptions = false', () => {
        const compiler = createCompiler(configClientBasic, configServerBasic);
        const spy = jest.spyOn(compiler, 'watch');

        webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: false,
        });

        expect(spy.mock.calls).toHaveLength(0);
        expect(compiler.isCompiling()).toBe(false);
    });

    it('should set headers as specified in options.headers', () => {
        const app = express();
        const compiler = createCompiler(configClientBasic, configServerBasic);

        app.use(webpackIsomorphicDevMiddleware(compiler, {
            watchOptions: { report: false },
            headers: { 'X-FOO': 'bar' },
        }));

        return request(app)
        .get('/client.js')
        .expect(200)
        .expect('X-FOO', 'bar')
        .expect(/Hello!/);
    });
});
