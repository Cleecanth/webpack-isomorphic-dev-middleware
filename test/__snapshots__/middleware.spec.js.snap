// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware should not re-require the server file if it has a runtime error 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <title>webpack-isomorphic-dev-middleware error</title>
        <style>
        body {
            background: #262626;
            color: #e6e6e6;
            line-height: 1.2;
            font-family: Menlo, Consolas, monospace;
            font-size: 13px;
            white-space: pre;
            margin: 10px;
        }

        .ansi-black-fg, .ansi-bright-black-fg { color: #6d7891; }
        .ansi-white-fg, .ansi-bright-white-fg { color: #fff; }
        .ansi-red-fg, .ansi-bright-red-fg { color: #e36049; }
        .ansi-green-fg, .ansi-bright-green-fg { color: #9eb567; }
        .ansi-yellow-fg, .ansi-bright-yellow-fg { color: #ffd080; }
        .ansi-blue-fg, .ansi-bright-blue-fg { color: #7cafc2; }
        .ansi-magenta-fg, .ansi-bright-magenta-fg { color: #d6add5; }
        .ansi-cyan-fg, .ansi-bright-cyan-fg { color: #c3c2ef; }

        [class$=\\"-bg\\"] { padding: 2px 4px; border-radius: 2px; }
        .ansi-black-bg, .ansi-bright-black-bg { background-color: #6d7891; color: #fff; }
        .ansi-white-bg, .ansi-bright-white-bg { background-color: #fff; color: #464646; }
        .ansi-red-bg, .ansi-bright-red-bg { background-color: #e36049; color: #fff; }
        .ansi-green-bg, .ansi-bright-green-bg { background-color: #9eb567; color: #464646; }
        .ansi-yellow-bg, .ansi-bright-yellow-bg { background-color: #ffd080; }
        .ansi-blue-bg, .ansi-bright-blue-bg { background-color: #7cafc2; color: #fff; }
        .ansi-magenta-bg, .ansi-bright-magenta-bg { background-color: #d6add5; color: #464646; }
        .ansi-cyan-bg, .ansi-bright-cyan-bg { background-color: #c3c2ef; color: #464646; }

        .ansi-reverse { background-color: #fff; color: #464646; }
        .ansi-dim { opacity: 0.7; }
        .ansi-italic { font-style: italic; }
        .ansi-underline { text-decoration: underline; }
        </style>
    </head>
    <body><span class=\\"ansi-dim\\">ReferenceError: </span>foo is not defined
[stack]

The error above was thrown while trying to load the built server file:
test/tmp/server.js</body>
</html>"
`;

exports[`middleware should not re-require the server file if it has a runtime error 2`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <title>webpack-isomorphic-dev-middleware error</title>
        <style>
        body {
            background: #262626;
            color: #e6e6e6;
            line-height: 1.2;
            font-family: Menlo, Consolas, monospace;
            font-size: 13px;
            white-space: pre;
            margin: 10px;
        }

        .ansi-black-fg, .ansi-bright-black-fg { color: #6d7891; }
        .ansi-white-fg, .ansi-bright-white-fg { color: #fff; }
        .ansi-red-fg, .ansi-bright-red-fg { color: #e36049; }
        .ansi-green-fg, .ansi-bright-green-fg { color: #9eb567; }
        .ansi-yellow-fg, .ansi-bright-yellow-fg { color: #ffd080; }
        .ansi-blue-fg, .ansi-bright-blue-fg { color: #7cafc2; }
        .ansi-magenta-fg, .ansi-bright-magenta-fg { color: #d6add5; }
        .ansi-cyan-fg, .ansi-bright-cyan-fg { color: #c3c2ef; }

        [class$=\\"-bg\\"] { padding: 2px 4px; border-radius: 2px; }
        .ansi-black-bg, .ansi-bright-black-bg { background-color: #6d7891; color: #fff; }
        .ansi-white-bg, .ansi-bright-white-bg { background-color: #fff; color: #464646; }
        .ansi-red-bg, .ansi-bright-red-bg { background-color: #e36049; color: #fff; }
        .ansi-green-bg, .ansi-bright-green-bg { background-color: #9eb567; color: #464646; }
        .ansi-yellow-bg, .ansi-bright-yellow-bg { background-color: #ffd080; }
        .ansi-blue-bg, .ansi-bright-blue-bg { background-color: #7cafc2; color: #fff; }
        .ansi-magenta-bg, .ansi-bright-magenta-bg { background-color: #d6add5; color: #464646; }
        .ansi-cyan-bg, .ansi-bright-cyan-bg { background-color: #c3c2ef; color: #464646; }

        .ansi-reverse { background-color: #fff; color: #464646; }
        .ansi-dim { opacity: 0.7; }
        .ansi-italic { font-style: italic; }
        .ansi-underline { text-decoration: underline; }
        </style>
    </head>
    <body><span class=\\"ansi-dim\\">ReferenceError: </span>foo is not defined
[stack]

The error above was thrown while trying to load the built server file:
test/tmp/server.js</body>
</html>"
`;

exports[`middleware should render error if an error occurred while reading the server file 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <title>webpack-isomorphic-dev-middleware error</title>
        <style>
        body {
            background: #262626;
            color: #e6e6e6;
            line-height: 1.2;
            font-family: Menlo, Consolas, monospace;
            font-size: 13px;
            white-space: pre;
            margin: 10px;
        }

        .ansi-black-fg, .ansi-bright-black-fg { color: #6d7891; }
        .ansi-white-fg, .ansi-bright-white-fg { color: #fff; }
        .ansi-red-fg, .ansi-bright-red-fg { color: #e36049; }
        .ansi-green-fg, .ansi-bright-green-fg { color: #9eb567; }
        .ansi-yellow-fg, .ansi-bright-yellow-fg { color: #ffd080; }
        .ansi-blue-fg, .ansi-bright-blue-fg { color: #7cafc2; }
        .ansi-magenta-fg, .ansi-bright-magenta-fg { color: #d6add5; }
        .ansi-cyan-fg, .ansi-bright-cyan-fg { color: #c3c2ef; }

        [class$=\\"-bg\\"] { padding: 2px 4px; border-radius: 2px; }
        .ansi-black-bg, .ansi-bright-black-bg { background-color: #6d7891; color: #fff; }
        .ansi-white-bg, .ansi-bright-white-bg { background-color: #fff; color: #464646; }
        .ansi-red-bg, .ansi-bright-red-bg { background-color: #e36049; color: #fff; }
        .ansi-green-bg, .ansi-bright-green-bg { background-color: #9eb567; color: #464646; }
        .ansi-yellow-bg, .ansi-bright-yellow-bg { background-color: #ffd080; }
        .ansi-blue-bg, .ansi-bright-blue-bg { background-color: #7cafc2; color: #fff; }
        .ansi-magenta-bg, .ansi-bright-magenta-bg { background-color: #d6add5; color: #464646; }
        .ansi-cyan-bg, .ansi-bright-cyan-bg { background-color: #c3c2ef; color: #464646; }

        .ansi-reverse { background-color: #fff; color: #464646; }
        .ansi-dim { opacity: 0.7; }
        .ansi-italic { font-style: italic; }
        .ansi-underline { text-decoration: underline; }
        </style>
    </head>
    <body>Failed to read file
[stack]

The error above was thrown while trying to load the built server file:
test/tmp/server.js</body>
</html>"
`;

exports[`middleware should render error if the server file has a runtime error 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <title>webpack-isomorphic-dev-middleware error</title>
        <style>
        body {
            background: #262626;
            color: #e6e6e6;
            line-height: 1.2;
            font-family: Menlo, Consolas, monospace;
            font-size: 13px;
            white-space: pre;
            margin: 10px;
        }

        .ansi-black-fg, .ansi-bright-black-fg { color: #6d7891; }
        .ansi-white-fg, .ansi-bright-white-fg { color: #fff; }
        .ansi-red-fg, .ansi-bright-red-fg { color: #e36049; }
        .ansi-green-fg, .ansi-bright-green-fg { color: #9eb567; }
        .ansi-yellow-fg, .ansi-bright-yellow-fg { color: #ffd080; }
        .ansi-blue-fg, .ansi-bright-blue-fg { color: #7cafc2; }
        .ansi-magenta-fg, .ansi-bright-magenta-fg { color: #d6add5; }
        .ansi-cyan-fg, .ansi-bright-cyan-fg { color: #c3c2ef; }

        [class$=\\"-bg\\"] { padding: 2px 4px; border-radius: 2px; }
        .ansi-black-bg, .ansi-bright-black-bg { background-color: #6d7891; color: #fff; }
        .ansi-white-bg, .ansi-bright-white-bg { background-color: #fff; color: #464646; }
        .ansi-red-bg, .ansi-bright-red-bg { background-color: #e36049; color: #fff; }
        .ansi-green-bg, .ansi-bright-green-bg { background-color: #9eb567; color: #464646; }
        .ansi-yellow-bg, .ansi-bright-yellow-bg { background-color: #ffd080; }
        .ansi-blue-bg, .ansi-bright-blue-bg { background-color: #7cafc2; color: #fff; }
        .ansi-magenta-bg, .ansi-bright-magenta-bg { background-color: #d6add5; color: #464646; }
        .ansi-cyan-bg, .ansi-bright-cyan-bg { background-color: #c3c2ef; color: #464646; }

        .ansi-reverse { background-color: #fff; color: #464646; }
        .ansi-dim { opacity: 0.7; }
        .ansi-italic { font-style: italic; }
        .ansi-underline { text-decoration: underline; }
        </style>
    </head>
    <body><span class=\\"ansi-dim\\">ReferenceError: </span>foo is not defined
[stack]

The error above was thrown while trying to load the built server file:
test/tmp/server.js</body>
</html>"
`;

exports[`middleware should wait for a failed compilation and render the webpack stats 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
    <head>
        <meta charset=\\"utf-8\\">
        <title>webpack-isomorphic-dev-middleware error</title>
        <style>
        body {
            background: #262626;
            color: #e6e6e6;
            line-height: 1.2;
            font-family: Menlo, Consolas, monospace;
            font-size: 13px;
            white-space: pre;
            margin: 10px;
        }

        .ansi-black-fg, .ansi-bright-black-fg { color: #6d7891; }
        .ansi-white-fg, .ansi-bright-white-fg { color: #fff; }
        .ansi-red-fg, .ansi-bright-red-fg { color: #e36049; }
        .ansi-green-fg, .ansi-bright-green-fg { color: #9eb567; }
        .ansi-yellow-fg, .ansi-bright-yellow-fg { color: #ffd080; }
        .ansi-blue-fg, .ansi-bright-blue-fg { color: #7cafc2; }
        .ansi-magenta-fg, .ansi-bright-magenta-fg { color: #d6add5; }
        .ansi-cyan-fg, .ansi-bright-cyan-fg { color: #c3c2ef; }

        [class$=\\"-bg\\"] { padding: 2px 4px; border-radius: 2px; }
        .ansi-black-bg, .ansi-bright-black-bg { background-color: #6d7891; color: #fff; }
        .ansi-white-bg, .ansi-bright-white-bg { background-color: #fff; color: #464646; }
        .ansi-red-bg, .ansi-bright-red-bg { background-color: #e36049; color: #fff; }
        .ansi-green-bg, .ansi-bright-green-bg { background-color: #9eb567; color: #464646; }
        .ansi-yellow-bg, .ansi-bright-yellow-bg { background-color: #ffd080; }
        .ansi-blue-bg, .ansi-bright-blue-bg { background-color: #7cafc2; color: #fff; }
        .ansi-magenta-bg, .ansi-bright-magenta-bg { background-color: #d6add5; color: #464646; }
        .ansi-cyan-bg, .ansi-bright-cyan-bg { background-color: #c3c2ef; color: #464646; }

        .ansi-reverse { background-color: #fff; color: #464646; }
        .ansi-dim { opacity: 0.7; }
        .ansi-italic { font-style: italic; }
        .ansi-underline { text-decoration: underline; }
        </style>
    </head>
    <body>Webpack client-side compilation failed

<span class=\\"ansi-bright-red-fg\\">ERROR in ./test/configs/files/syntax-error.js
Module parse failed: test/configs/files/syntax-error.js Unexpected token (4:0)
You may need an appropriate loader to handle this file type.
| 
| console.log(&#39;Hello!&#39;
| </span></body>
</html>"
`;
